### ISSUE TITLE
ListSwaps not picking up failed swap state

### ISSUE THREAD

**Issue by guggero** [lightninglabs/loop#354]
References https://github.com/lightninglabs/lightning-terminal/issues/200.

It seems like the swap state sometimes isn't updated in the result of `ListSwaps` (which the UI uses) if it fails.

The log file (https://github.com/lightninglabs/lightning-terminal/files/6103144/litd.log) clearly showed this:
```
f99673 Server update: FAILED_OFF_CHAIN_TIMEOUT received, timestamp: 2021-03-07 16:15:25.601380392 +0000 UTC
```

But the response was only updated after a restart of loopd/LiT.

**Comment by carlaKC** [comment]
Struggling to reproduce this on my end. 

From the logs it looks like the swap is getting stuck when we wait to [finalize payments](https://github.com/lightninglabs/loop/blob/master/loopout.go#L338), which is strange because it's getting a prepay failed notification [here](https://github.com/lightninglabs/loop/blob/master/loopout.go#L700) so should just skip through the final wait and mark the swap as [complete](https://github.com/lightninglabs/loop/blob/master/loopout.go#L371).

Any chance we have logs from after the restart which fixed the issue? I'd be interested to see how it resumed and resolved itself.

**Comment by guggero** [comment]
My suspicion is that this perhaps has something to do with restarting `loopd` while the swap is still on its way?
Haven't looked at the log in much detail yet but when I scanned it the first time I noticed quite a few restarts.

I'll ask for the new logs.

**Comment by carlaKC** [comment]
`f99673` gets stuck right at the end of the logs unfortunately :')

**Comment by guggero** [comment]
Here's the latest log: https://github.com/lightninglabs/lightning-terminal/files/6123518/litd.log.

Looks like it was indeed only picked up on restart:

```
2021-03-10 13:55:47.310 [INF] LOOP: f99673 Loop out swap state: Initiated
2021-03-10 13:55:47.311 [INF] LOOP: f99673 Sending swap payment lnbc291796800n1psyfa6jpp5lxt88h3srs2drcjvuumq4t9v988n3yp62zx67ydjmwmsnscaj4kqdq8wdmkzuqcqzxgxq97zvuqrzjqgwf02g2gy0l9vgdc25wxt0z72wjlfyagxlmk54ag9hyvrdsw37s6zs6syqqrmcqqyqqqqqqqqqqqqqq9qrzjqgwf02g2gy0l9vgdc25wxt0z72wjlfyagxlmk54ag9hyvrdsw37s6zs79qqqz4cqqqqqqqqqqqqqqqqq9qsp5lgh9xy8quj902puj5xa873lxfalupjen529ptm2m8htsasxz5veq9qyyssqm0d8e9p32mw8mhurjm5lea4rt9mc2qr0vfv0fweap99872sstv5py5kn38er2t5kw3j3dxpe0g8jdpmm89x367f9fl8qv2tkre3tm8sqp49aq2
2021-03-10 13:55:47.311 [INF] LOOP: f99673 Sending prepayment lnbc300u1psyfa6jpp5ahnctmuzjl7wf4l620yfapfp3u2dzrhzpk4ulewr8m0e7dfqxcksdq2wpex2urp0ycqzxgxq97zvuqrzjqgwf02g2gy0l9vgdc25wxt0z72wjlfyagxlmk54ag9hyvrdsw37s6zs6syqqrmcqqyqqqqqqqqqqqqqq9qrzjqgwf02g2gy0l9vgdc25wxt0z72wjlfyagxlmk54ag9hyvrdsw37s6z34hyqqdlcqqyqqqqqqqqqqqqqq9qrzjqgwf02g2gy0l9vgdc25wxt0z72wjlfyagxlmk54ag9hyvrdsw37s6zs79qqqz4cqqqqqqqqqqqqqqqqq9qsp5yvz4pma78qtd2kaqujs48ycxqner38l525sustpxt8jsc2jv0k0q9qyyssq874e3uhzzxqju0q7j9xeankyzzf22kvddum66z326kewfe0nncv9hagutgtx42mf57pn02jynzp056a6jsemg2gtghfdg76ajpmzgqcqzdjgxa
2021-03-10 13:55:47.311 [INF] LOOP: f99673 Register 1 conf ntfn for swap script on chain (hh=673575)
2021-03-10 13:55:47.323 [INF] LOOP: f99673 Checking preimage reveal height 673655 exceeded (height 674010)
2021-03-10 13:55:47.323 [INF] LOOP: f99673 Max preimage reveal height 673655 exceeded (height 674010)
2021-03-10 13:55:47.324 [INF] LOOP: f99673 Wait for server pulling off-chain payment(s)
2021-03-10 13:55:47.434 [INF] LOOP: f99673 Payment ede785ef8297fce4d7fa53c89e85218f14d10ee20dabcfe5c33edf9f3520362d: state=IN_FLIGHT, inflight_htlcs=0, inflight_amt=0 mSAT
2021-03-10 13:55:47.565 [INF] LOOP: f99673 Payment f99673de301c14d1e24ce7360aacac29cf38903a508daf11b2dbb709c31d956c: state=IN_FLIGHT, inflight_htlcs=2, inflight_amt=8076517411 mSAT
2021-03-10 13:55:47.704 [INF] LOOP: f99673 Server update: FAILED_OFF_CHAIN_TIMEOUT received, timestamp: 2021-03-07 16:15:25.642223884 +0000 UTC
2021-03-10 13:55:47.705 [INF] LOOP: f99673 swap subscription: server finished serving updates
2021-03-10 13:55:51.041 [INF] LOOP: f99673 Payment ede785ef8297fce4d7fa53c89e85218f14d10ee20dabcfe5c33edf9f3520362d: state=FAILED
2021-03-10 13:55:51.041 [INF] LOOP: f99673 Prepayment failed: payment failed
2021-03-10 13:55:55.243 [INF] LOOP: f99673 Payment f99673de301c14d1e24ce7360aacac29cf38903a508daf11b2dbb709c31d956c: state=IN_FLIGHT, inflight_htlcs=1, inflight_amt=3907994079 mSAT
2021-03-10 13:55:55.435 [INF] LOOP: f99673 Payment f99673de301c14d1e24ce7360aacac29cf38903a508daf11b2dbb709c31d956c: state=FAILED
2021-03-10 13:55:55.436 [INF] LOOP: f99673 Swap payment failed: payment failed
2021-03-10 13:55:55.436 [INF] LOOP: f99673 Swap completed: FailTimeout (final cost: server 0 BTC, onchain 0 BTC, offchain 0 BTC)
2021-03-10 13:55:55.443 [INF] LOOP: f99673 Loop out swap state: FailTimeout
```

**Comment by lukaszsamson** [comment]
I'm seeing a similar issue. In my case the prepayment failed and loop out is stuck in initiating state even long after preimage reveal height is exceeded. It may be related to https://github.com/lightningnetwork/lnd/issues/4788 as I can see 76e0570ea031be6f584b69fa2f67f2c8ef320beee5295be24671db22ef0ee7fb is IN_FLIGHT but in lnd logs there is a message that it failed.

litd logs after start
```
2021-03-20 11:55:41.784 [INF] LOOP: 76e057 Payment 76e0570ea031be6f584b69fa2f67f2c8ef320beee5295be24671db22ef0ee7fb: state=IN_FLIGHT, inflight_htlcs=3, inflight_amt=1215718526 mSAT
2021-03-20 11:55:43.982 [INF] LOOP: 76e057 Payment 76e0570ea031be6f584b69fa2f67f2c8ef320beee5295be24671db22ef0ee7fb: state=IN_FLIGHT, inflight_htlcs=4, inflight_amt=1619134263 mSAT
2021-03-20 11:55:44.162 [INF] LOOP: 76e057 Payment 76e0570ea031be6f584b69fa2f67f2c8ef320beee5295be24671db22ef0ee7fb: state=IN_FLIGHT, inflight_htlcs=3, inflight_amt=1212314985 mSAT
2021-03-20 11:55:44.686 [INF] LOOP: 76e057 Payment 76e0570ea031be6f584b69fa2f67f2c8ef320beee5295be24671db22ef0ee7fb: state=IN_FLIGHT, inflight_htlcs=3, inflight_amt=1212314985 mSAT
2021/03/20 12:02:32 http: TLS handshake error from 127.0.0.1:64608: remote error: tls: unknown certificate
2021/03/20 12:02:35 http: TLS handshake error from 127.0.0.1:64623: remote error: tls: unknown certificate
```

corresponding lnd logs
```
2021-03-20 11:55:43.704 [WRN] CRTR: Attempt 6239 for payment 76e0570ea031be6f584b69fa2f67f2c8ef320beee5295be24671db22ef0ee7fb failed: IncorrectOrUnknownPaymentDetails(amt=406819278 mSAT, height=675453)@7
2021-03-20 11:55:44.289 [INF] CRTR: Payment 76e0570ea031be6f584b69fa2f67f2c8ef320beee5295be24671db22ef0ee7fb failed: final_outcome=incorrect_payment_details, raw_err=IncorrectOrUnknownPaymentDetails(amt=406819278 mSAT, height=675453)@7
2021-03-20 11:55:44.583 [ERR] CRTR: Payment with hash 76e0570ea031be6f584b69fa2f67f2c8ef320beee5295be24671db22ef0ee7fb failed: payment has reached terminal condition
```

The payment failed but `lncli listpayments --include_incomplete` still lists it as IN_FLIGHT

litd after restart
```
2021-03-20 13:08:47.394 [INF] LOOP: Resuming loop out swap 76e0570ea031be6f584b69fa2f67f2c8ef320beee5295be24671db22ef0ee7fb
2021-03-20 13:08:47.394 [INF] LOOP: Wait for first block ntfn
2021-03-20 13:08:47.394 [INF] LOOP: 76e057 Htlc address: bc1qljpf0zc4lwnrqus85jchdkheen4dg3q0kf4wd7tlva52w483qadq9m5q6j
2021-03-20 13:08:47.402 [INF] LOOP: Starting event loop at height 675460
2021-03-20 13:08:47.402 [INF] LOOP: 76e057 Loop out swap state: Initiated
2021-03-20 13:08:47.402 [INF] LOOP: 76e057 Sending swap payment lnbc20225500n1ps9970hpp5wms9wr4qxxlx7kztd8az7eljerhnyzlwu554hcjxw8dj9mcwulasdq8wdmkzuqcqzxgxq97zvuqrzjqgwf02g2gy0l9vgdc25wxt0z72wjlfyagxlmk54ag9hyvrdsw37s6zs6syqqrmcqqyqqqqqqqqqqqqqq9qrzjqgwf02g2gy0l9vgdc25wxt0z72wjlfyagxlmk54ag9hyvrdsw37s6z34hyqqdlcqqyqqqqqqqqqqqqqq9qsp5e9c4xdxxh0v0k7yg6yxvvzgpvyctzlqjt59hhdluyldt0hhg475q9qyyssqhv3h7ehke0sv7h758e439rtyftxhgg4cyah9kva4q86d0zedat8qaxl5meg72e32f2v4dqetmfrqd2hent47h6xrjhkcl896ytyg0lqqp4zghy
2021-03-20 13:08:47.402 [INF] LOOP: 76e057 Sending prepayment lnbc300u1ps9970hpp5kv8eqfc7kkqr4e83dpps963cv7383qwpdx6d5geq8pf20gysqcdsdq2wpex2urp0ycqzxgxq97zvuqrzjqgwf02g2gy0l9vgdc25wxt0z72wjlfyagxlmk54ag9hyvrdsw37s6zs6syqqrmcqqyqqqqqqqqqqqqqq9qrzjqgwf02g2gy0l9vgdc25wxt0z72wjlfyagxlmk54ag9hyvrdsw37s6z34hyqqdlcqqyqqqqqqqqqqqqqq9qrzjqgwf02g2gy0l9vgdc25wxt0z72wjlfyagxlmk54ag9hyvrdsw37s6zs79qqqz4cqqqqqqqqqqqqqqqqq9qsp5rdexp6ka35xl4pd9gsya53cc30a298a0uc0rs4a9pdg7j5pvkqfq9qyyssq49zjtpwt5zr3a4fm2vg4pdugfa4la6hymkjqw3e9hqcuut3xc2x99m5twfsuxj3pps2fc8wy5uste7gjn3x6qm04fdx04pwsctldzqgplhthyy
2021-03-20 13:08:47.402 [INF] LOOP: 76e057 Register 1 conf ntfn for swap script on chain (hh=675130)
2021-03-20 13:08:47.403 [INF] LOOP: 76e057 Checking preimage reveal height 675210 exceeded (height 675460)
2021-03-20 13:08:47.403 [INF] LOOP: 76e057 Max preimage reveal height 675210 exceeded (height 675460)
2021-03-20 13:08:47.403 [INF] LOOP: 76e057 Wait for server pulling off-chain payment(s)
2021-03-20 13:08:47.493 [INF] LLMD: Auction server address: pool.lightning.finance:12010
2021-03-20 13:08:47.519 [INF] CLDB: Checking for schema update: latest_version=1, db_version=1
2021-03-20 13:08:47.520 [INF] RPCS: Starting trader server
2021-03-20 13:08:47.600 [INF] LOOP: 76e057 Payment b30f90271eb5803ae4f1684302ea3867a27881c169b4da23203852a7a090061b: state=IN_FLIGHT, inflight_htlcs=0, inflight_amt=0 mSAT
2021-03-20 13:08:47.779 [INF] RPCS: Connected to lnd node *** with pubkey ***
2021-03-20 13:08:47.801 [INF] LOOP: 76e057 Payment b30f90271eb5803ae4f1684302ea3867a27881c169b4da23203852a7a090061b: state=IN_FLIGHT, inflight_htlcs=1, inflight_amt=30000000 mSAT
2021-03-20 13:08:47.907 [INF] LOOP: 76e057 Payment 76e0570ea031be6f584b69fa2f67f2c8ef320beee5295be24671db22ef0ee7fb: state=IN_FLIGHT, inflight_htlcs=3, inflight_amt=1212314985 mSAT
2021-03-20 13:08:47.977 [INF] RPCS: Trader server is now active
2021-03-20 13:08:48.800 [INF] LOOP: 76e057 Server update: FAILED_OFF_CHAIN_TIMEOUT received, timestamp: 2021-03-18 08:18:35.522860161 +0100 CET
2021-03-20 13:08:48.800 [INF] LOOP: 76e057 swap subscription: server finished serving updates
2021-03-20 13:08:50.373 [INF] LOOP: 76e057 Payment b30f90271eb5803ae4f1684302ea3867a27881c169b4da23203852a7a090061b: state=IN_FLIGHT, inflight_htlcs=0, inflight_amt=0 mSAT
2021-03-20 13:08:50.579 [INF] LOOP: 76e057 Payment b30f90271eb5803ae4f1684302ea3867a27881c169b4da23203852a7a090061b: state=IN_FLIGHT, inflight_htlcs=1, inflight_amt=30000000 mSAT
2021-03-20 13:09:25.346 [INF] LOOP: 76e057 Payment b30f90271eb5803ae4f1684302ea3867a27881c169b4da23203852a7a090061b: state=IN_FLIGHT, inflight_htlcs=0, inflight_amt=0 mSAT
2021-03-20 13:09:25.702 [INF] LOOP: 76e057 Payment b30f90271eb5803ae4f1684302ea3867a27881c169b4da23203852a7a090061b: state=FAILED
2021-03-20 13:09:25.702 [INF] LOOP: 76e057 Prepayment failed: payment failed
```

lnd after restart correctly sets payment as failed
```
2021-03-20 16:44:56.570 [ERR] CRTR: Resuming payment with hash 76e0570ea031be6f584b69fa2f67f2c8ef320beee5295be24671db22ef0ee7fb failed: incorrect_payment_details.
```

Another litd restart and the cycle repeats.

**Comment by carlaKC** [comment]
Thanks for the report @lukaszsamson!

Looks like both of these stuck swaps are due to the swap payment being stuck in flight. This can happen in the following case:
- Loop client creates a swap
- Prepay payment routes to server and is accepted
- Swap payment is busy trying to find a route to the server
- Server times the swap out, and cancels the prepay and swap invoices
- We hit the [race](https://github.com/lightningnetwork/lnd/issues/4788#issuecomment-804093718) in lnd because we get a fatal error (inv cancelled) while we're still dispatching some shards
- Swap payment gets stuck and loop will get stuck waiting for a final state on the payment

Seems like a restart of litd sometimes helps because lnd may not hit the same race condition (although it may sometimes get stuck again). 

A fix for #4788 in lnd should address these issues, since the swap will pickup the payment failure and resolve once and for all. In the meantime, all of your funds are back in your control, so this is mostly an aesthetic issue (although an annoying one!). I'm going to close this issue for now, pending fix in lnd.
